### DISCUSSION TITLE
How do i use connection.on('.....')

### DISCUSSION THREAD

**Discussion by FreshImmuc** [QuorumDMS/ftp-srv#325] [Q&A] [answered]
How (where) do i use this:
```js
connection.on('STOR', (error, fileName) => {
        // ...
});
```

Thanks in advance

**Comment by matt-forster** [comment, marked as answer]
This would be a _side effect_ of storing a file. 

The file has already been stored, and the STOR handler would have already finished. You can see the emit here, right in the default STOR commend: https://github.com/QuorumDMS/ftp-srv/blob/main/src/commands/registration/stor.js#L53

So you would listen to this if you want to log out specifically the file that was stored, maybe send a metric to your observability platform, etc.

**Reply by FreshImmuc** [reply]
Okay, do you have an example of it? Maybe in the basic server example from the documentation
![image](https://user-images.githubusercontent.com/65733898/175352554-d4f3c249-dbfe-46a9-8e7d-4e84345bb035.png)
Here in text:
```js
// Quick start, create an active ftp server.
const FtpSrv = require('ftp-srv');

const port=21;
const ftpServer = new FtpSrv({
    url: "ftp://0.0.0.0:" + port,
    anonymous: true
});

ftpServer.on('login', (data, resolve, reject) => { 
    if(data.username === 'anonymous' && data.password === 'anonymous'){
        return resolve({ root:"/" });    
    }
    return reject(new errors.GeneralError('Invalid username or password', 401));
});

ftpServer.listen().then(() => { 
    console.log('Ftp server is starting...')
});
```

Where would i insert/use this?

**Reply by matt-forster** [reply]
The event is emitted on the connection - which is available in a few different contexts - mainly, the login event, and potentially the filesystem constructor.

https://github.com/QuorumDMS/ftp-srv/blob/1af62a7c4f3227b3707037813a0c7fd630c4147f/src/fs.js#L13

If you return a filesystem constructor from the login promise (as below), it will be used, otherwise, the default will be used:

https://github.com/QuorumDMS/ftp-srv/blob/1af62a7c4f3227b3707037813a0c7fd630c4147f/src/connection.js#L91

```js
// Quick start, create an active ftp server.
const FtpSrv = require('ftp-srv');

const port=21;
const ftpServer = new FtpSrv({
    url: "ftp://0.0.0.0:" + port,
    anonymous: true
});

ftpServer.on('login', ({ connection, username, password }, resolve, reject) => {
   // If creating a custom filesystem
   // You'll be able to use the connection.on('STOR', ...) event inside the filesystem wherever you want.
   const fs = new CustomFilesystem({connection});
 
    if(data.username === 'anonymous' && data.password === 'anonymous') {
       // If using the default, or you have other side effects
       connection.on('STOR', ...);

        // if using a custom filesystem
        return resolve({ fs });    
    }

    return reject(new errors.GeneralError('Invalid username or password', 401));
});

ftpServer.listen().then(() => { 
    console.log('Ftp server is starting...')
});
```

**Reply by FreshImmuc** [reply]
Well now it says 
```js
connection.on('STOR', (error, fileName) => {
^

ReferenceError: connection is not defined
```

**Reply by matt-forster** [reply]
Is connection defined in the scope that you are using that expression?

**Reply by matt-forster** [reply]
Here is the login event: https://github.com/QuorumDMS/ftp-srv/blob/0b55b3f79d894f47f92ee3dc9c884502b1a93cd3/src/connection.js#L84

**Reply by matt-forster** [reply]
@FreshImmuc Here is a documentation clarification: https://github.com/QuorumDMS/ftp-srv/pull/326

**Reply by matt-forster** [reply]
You'll see the full login event here: https://github.com/QuorumDMS/ftp-srv#login

**Reply by FreshImmuc** [reply]
Ah okay yeah i fixed some stuff and in the end my only "error" was that i didn't wrote "data" as an object in {}, i just put 
```
data
``` 
instead of
```js
{ connection, username, password }
```
and it works now, thank you!
